File: Java/Intervals/MergeIntervals.java
package Intervals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class MergeIntervals {
    // Leetcode 56: https://leetcode.com/problems/merge-intervals/description/
    // Submission Detail: https://leetcode.com/problems/merge-intervals/submissions/1626710995

    // Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
    // Output: [[1,6],[8,10],[15,18]]
    // Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].
    //
    // Input: intervals = [[1,4],[4,5]]
    // Output: [[1,5]]
    // Explanation: Intervals [1,4] and [4,5] are considered overlapping.

    // Tx = O(nlogn)
    // Sx = O(n)
    public int[][] merge(int[][] intervals) {
        List<int[]> mergedIntervals = new ArrayList<>();
        Arrays.sort(intervals, Comparator.comparingInt(interval -> interval[0])); // sort by start time
        mergedIntervals.add(intervals[0]);

        for(int index=1; index<intervals.length; index++) {
            int size = mergedIntervals.size();
            if(intervals[index][0] <= mergedIntervals.get(size-1)[1]) {
                mergedIntervals.get(size-1)[1] = Math.max(mergedIntervals.get(size-1)[1], intervals[index][1]);
            } else {
                mergedIntervals.add(intervals[index]);
            }
        }

        return mergedIntervals.toArray(new int[mergedIntervals.size()][]);
    }
}

File: C#/Leetcode/Patterns/Intervals/MergeIntervals.cs
﻿using sys = System;
using System.Collections.Generic;
using System.Linq;
using LeetcodeSolutions.DataStructures;

namespace LeetcodeSolutions.Patterns.Intervals
{
    // Leetcode 56 - https://leetcode.com/problems/merge-intervals/
    // Submission Detail - https://leetcode.com/submissions/detail/128324956/

    public class MergeIntervals
    {
        //[[4,5],[2,4],[4,6],[3,4],[0,0],[1,1],[3,5],[2,2]]
        //[[1,4],[2,3]]
        //[[2,3],[5,5],[2,2],[3,4],[3,4]]
        //[[4,4],[3,5],[2,3],[1,1],[3,3],[1,3],[2,2],[0,0],[5,5],[0,0],[4,6]] - sorted [[0,0],[0,0],[1,1],[1,3],[2,2],[2,3],[3,3],[3,5],[4,4],[4,6],[5,5]]

        // Runtime = 655 ms
        // Tx = O(nlogn)
        // Sx = O(n) for output list
        public IList<Interval> Merge(List<Interval> intervals)
        {
            if (intervals == null)
                return intervals;

            // Sort the array on the start time.
            intervals.Sort((i1, i2) => i1.start.CompareTo(i2.start));

            IList<Interval> mergedIntervals = new List<Interval>{
                intervals[0]
            };

            foreach (var currentInterval in intervals)
            {
                Interval lastMergedInterval = mergedIntervals.Last();

                if (lastMergedInterval.end >= currentInterval.start)
                    lastMergedInterval.end = sys.Math.Max(currentInterval.end, lastMergedInterval.end);
                else
                    // Last and current intervals do not overlap. So, add current interval at last in the list.
                    mergedIntervals.Add(currentInterval);
            }

            return mergedIntervals;
        }
    }
}
